### ISSUE TITLE
Provide access to raw packet data via advertisment.raw_data 

### ISSUE THREAD

**Issue by deepcoder** [TheCellule/python-bleson#78]
In this post:
https://groups.google.com/g/python-bleson/c/xBG-kwWLkV0
you reference advertisement.data_raw. However when I try to get this raw data, all I get back in an empty object. How do you access the raw advertising string. I am using:
bleson in /usr/local/lib/python3.8/dist-packages (0.1.6)
on Ubuntu 20.04.1 LTS

This is the output:
```
BDAddress('F6:90:1E:17:5E:01') None
BDAddress('F6:90:1E:17:5E:01') None
BDAddress('52:F8:2D:75:69:61') None
BDAddress('22:5C:63:0B:97:C0') None
BDAddress('6E:46:35:6A:70:30') None
```

from sample basic_observer.py set to show advertisement.raw_data:

```
#!/usr/bin/env python3

import sys
from time import sleep
from bleson import logger, get_provider, Observer

# Get the wait time from the first script argument or default it to 10 seconds
WAIT_TIME = int(sys.argv[1]) if len(sys.argv)>1 else 10

#logger.set_level(10)

def on_advertisement(advertisement):
    print(advertisement.address, advertisement.raw_data)

adapter = get_provider().get_adapter()

observer = Observer(adapter)
observer.on_advertising_data = on_advertisement

while True :
    observer.start()
    sleep(WAIT_TIME)
    observer.stop()
    print(80 * "*")
```

Thanks for your work and help!

Dave Proffer

**Comment by WayneKeenan** [comment]
Hi, honestly I’m at a bit of a loss as to why you aren’t getting the data as Linux is the ‘strongest’ platform. But when I say Linux I’m only referring to Raspbian.  

My gut feeling is stop the Bluetooth daemon and retest. It’s just a hunch. I don’t have anything to hand to reproduce unfortunately.

**Comment by deepcoder** [comment]
Hi Wayne,
Thank you for your follow up.

I tried the following and no difference. Is this what you mean by 'stop the Bluetooth daemon'?

```
sudo systemctl stop bluetooth

sudo systemctl status bluetooth

● bluetooth.service - Bluetooth service
     Loaded: loaded (/lib/systemd/system/bluetooth.service; enabled; vendor preset: enabled)
     Active: inactive (dead) since Sun 2020-10-18 18:35:56 PDT; 48s ago
       Docs: man:bluetoothd(8)
    Process: 1490365 ExecStart=/usr/lib/bluetooth/bluetoothd (code=exited, status=0/SUCCESS)
   Main PID: 1490365 (code=exited, status=0/SUCCESS)
     Status: "Powering down"

Oct 18 18:35:56 ubuntu-macmini2012 bluetoothd[1490365]: Terminating
Oct 18 18:35:56 ubuntu-macmini2012 systemd[1]: Stopping Bluetooth service...
Oct 18 18:35:56 ubuntu-macmini2012 bluetoothd[1490365]: Endpoint unregistered: sender=:1.69 path=/MediaEndpoint/A2DPSink/sbc
Oct 18 18:35:56 ubuntu-macmini2012 bluetoothd[1490365]: Endpoint unregistered: sender=:1.69 path=/MediaEndpoint/A2DPSource/sbc
Oct 18 18:35:56 ubuntu-macmini2012 bluetoothd[1490365]: Endpoint unregistered: sender=:1.69 path=/MediaEndpoint/A2DPSink/sbc
Oct 18 18:35:56 ubuntu-macmini2012 bluetoothd[1490365]: Endpoint unregistered: sender=:1.69 path=/MediaEndpoint/A2DPSource/sbc
Oct 18 18:35:56 ubuntu-macmini2012 bluetoothd[1490365]: Stopping SDP server
Oct 18 18:35:56 ubuntu-macmini2012 bluetoothd[1490365]: Exit
Oct 18 18:35:56 ubuntu-macmini2012 systemd[1]: bluetooth.service: Succeeded.
Oct 18 18:35:56 ubuntu-macmini2012 systemd[1]: Stopped Bluetooth service.

sudo ./test.py

BDAddress('64:71:EE:D5:6B:3E') None
BDAddress('11:5B:9C:D6:03:34') None
BDAddress('EB:33:5C:1D:ED:E8') None
BDAddress('1F:93:D7:7C:08:24') None
BDAddress('C8:69:CD:31:FF:F5') None
BDAddress('C8:69:CD:31:FF:F5') None
BDAddress('5A:50:55:87:CC:F2') None
BDAddress('66:66:A3:06:91:C9') None
BDAddress('6E:46:35:6A:70:30') None
BDAddress('6E:46:35:6A:70:30') None

```

**Comment by Thrilleratplay** [comment]
@deepcoder Something I found to be weird is that raw_data is not set during parsing, you need to set it after the fact.

In the Linux adapter, before `on_advertising_data` callback is executed, the data is parsed by [AdvertisingDataConverters.from_hcipacket](https://github.com/TheCellule/python-bleson/blob/master/bleson/providers/linux/linux_adapter.py#L248), which breaks down the packet in [from_hcipayload](https://github.com/TheCellule/python-bleson/blob/master/bleson/core/hci/type_converters.py#L104) but does not set `raw_data`.  If you want to see this value, enable debugging but the parsed values are likely going to be the most use for you.

**Comment by deepcoder** [comment]
Hi @Thrilleratplay and @WayneKeenan , 

@WayneKeenan I tried sudo pip3 installing bleson on Ubuntu 18.04.5 LTS (Bionic Beaver) and Raspbian GNU/Linux 9 (stretch), both behave the same, aka no raw packet data.

@Thrilleratplay  thanks for your answer and information! I'm not sure I fully understand when you say ' you need to set it after the fact.' ? 

Is this something I can do in my code (if so how?), or are you saying that the BLESON module has a bug and does you set raw_data (and so BLESON needs a fix?)?

All that said, based on your further point to turn on logging to level DEBUG, I see that the raw advertising packet appears to be sent out in the logging stream. So a little ugly BASH code:

`
sudo ./test.py 2>&1 >/dev/null | grep -A 2 "LE Advertising Report" | cut -c 93-
`

yields the following output which I think is the raw data I am looking, TBD. FYI, I want to see the raw data just for my owned education, to get a better grasp on the various types of packets bluetooth sends around. Thank you for your help! Dave

output from above:

`
LE Advertising Report
`
`
Data: 01 02 01 e1 68 42 46 a3 69 1c 03 03 9f fe 17 16 9f fe 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 b5
`
`
POS : 00 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38
`
`
LE Advertising Report
`
`
Data: 01 03 01 24 08 7c d7 93 1f 0f 02 01 1a 0b ff 4c 00 09 06 03 85 c0 a8 02 f4 b7
POS : 00 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25
`
`
LE Advertising Report
`
`
Data: 01 00 01 77 f9 a5 df 9e 47 11 02 01 1a 02 0a 0c 0a ff 4c 00 10 05 01 14 79 9f 6c c1
POS : 00 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27
`
`
LE Advertising Report
`
`
Data: 01 00 01 c7 9f 0a c9 89 4a 12 02 01 1a 02 0a 0c 0b ff 4c 00 10 06 03 1d 49 b3 50 08 ba
POS : 00 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28
`

**Comment by Thrilleratplay** [comment]
@deepcoder Sorry, I should have been clearer.  One would assume, I did and apparently you did as well, that `raw_data` would be set while defining the advertisement object.  When searching for `raw_data` in this project, the only place I found where it is used is in the [eddystone beacon code](https://github.com/TheCellule/python-bleson/blob/master/bleson/beacons/eddystone.py#L44) where the value is manually set to an encoded url.  If I understand it correctly, `raw_data` is used when transmitting data but not observing.  I do not think it is a bug as it seems to be done one purpose; if you look at [from_advertisement in type converters](https://github.com/TheCellule/python-bleson/blob/master/bleson/core/hci/type_converters.py#L52) `raw_data` is the only thing returned in an advertisement returned when defined and searching for `from_advertisement ` it is only used when the interface is advertising.

The lines that are prefixed `Data:` are a hex string representation of the hci packet.  I think using `advertisement.__dir__` in your `on_advertisement` method will display all of the keys in the collections.  Iterating through this information will likely yield the parsed values you are ultimately looking for.
